File: cogant/examples/calculator_js/calculator.js
/**
 * Simple calculator with state machine (JavaScript port of
 * ``examples/control_positive/calculator/calculator.py``).
 *
 * Role contract (used by cross-language differential tests):
 *   - display, accumulator, operation, history, new_input: hidden state
 *   - input_digit, input_operation, equals, clear: actions
 *   - _execute_operation: hidden state transition
 *   - get_display, get_history: observations
 *   - assert_display, assert_history_length: preferences
 *
 * This file is intentionally kept free of external dependencies so the
 * tree-sitter JavaScript parser can walk it without a tool-chain.
 */

"use strict";

class Calculator {
  constructor() {
    // Hidden state
    this.display = "0";
    this.accumulator = 0;
    this.operation = null;
    this.history = [];
    this.new_input = true;
  }

  input_digit(digit) {
    // Action: input a digit (0-9).
    if (digit < 0 || digit > 9) {
      return this.display;
    }

    if (this.new_input) {
      this.display = String(digit);
      this.new_input = false;
    } else {
      this.display = this.display + String(digit);
    }

    this.history.push(`input_digit(${digit})`);
    return this.display;
  }

  input_operation(op) {
    // Action: input operation (+, -, *, /).
    const current_value = parseInt(this.display, 10);

    if (this.operation !== null) {
      // Execute pending operation.
      this.accumulator = this._execute_operation(
        this.accumulator,
        current_value,
        this.operation,
      );
      this.display = String(this.accumulator);
    } else {
      this.accumulator = current_value;
    }

    this.operation = op;
    this.new_input = true;
    this.history.push(`input_operation(${op})`);
    return this.display;
  }

  equals() {
    // Action: execute equals.
    if (this.operation === null) {
      return this.display;
    }

    const current_value = parseInt(this.display, 10);
    const result = this._execute_operation(
      this.accumulator,
      current_value,
      this.operation,
    );

    this.display = String(result);
    this.operation = null;
    this.accumulator = 0;
    this.new_input = true;
    this.history.push("equals()");
    return this.display;
  }

  clear() {
    // Action: clear the calculator.
    this.display = "0";
    this.accumulator = 0;
    this.operation = null;
    this.new_input = true;
    this.history.push("clear()");
    return this.display;
  }

  _execute_operation(left, right, op) {
    // Hidden state operation.
    if (op === "+") {
      return left + right;
    } else if (op === "-") {
      return left - right;
    } else if (op === "*") {
      return left * right;
    } else if (op === "/" && right !== 0) {
      return Math.trunc(left / right);
    }
    return left;
  }

  get_display() {
    // Observation: current display value.
    return this.display;
  }

  get_history() {
    // Observation: operation history.
    return this.history.slice();
  }

  assert_display(expected) {
    // Preference: assertion on display.
    return this.display === expected;
  }

  assert_history_length(max_length) {
    // Preference: constraint on history size.
    return this.history.length <= max_length;
  }
}

module.exports = { Calculator };

File: cogant/examples/control_positive/calculator/calculator.py
"""
Simple calculator with state machine.

Exercises: state machine patterns, operations as actions, display as observation,
history as hidden state, assertions as preferences.
"""

from typing import List, Optional


class Calculator:
    """A calculator with state machine semantics."""

    def __init__(self):
        # Hidden state
        self.display = "0"
        self.accumulator = 0
        self.operation = None
        self.history: List[str] = []
        self.new_input = True

    def input_digit(self, digit: int) -> str:
        """Action: input a digit (0-9)."""
        if digit < 0 or digit > 9:
            return self.display

        if self.new_input:
            self.display = str(digit)
            self.new_input = False
        else:
            self.display = self.display + str(digit)

        self.history.append(f"input_digit({digit})")
        return self.display

    def input_operation(self, op: str) -> str:
        """Action: input operation (+, -, *, /)."""
        current_value = int(self.display)

        if self.operation is not None:
            # Execute pending operation
            self.accumulator = self._execute_operation(
                self.accumulator, current_value, self.operation
            )
            self.display = str(self.accumulator)
        else:
            self.accumulator = current_value

        self.operation = op
        self.new_input = True
        self.history.append(f"input_operation({op})")
        return self.display

    def equals(self) -> str:
        """Action: execute equals."""
        if self.operation is None:
            return self.display

        current_value = int(self.display)
        result = self._execute_operation(self.accumulator, current_value, self.operation)

        self.display = str(result)
        self.operation = None
        self.accumulator = 0
        self.new_input = True
        self.history.append("equals()")
        return self.display

    def clear(self) -> str:
        """Action: clear the calculator."""
        self.display = "0"
        self.accumulator = 0
        self.operation = None
        self.new_input = True
        self.history.append("clear()")
        return self.display

    def _execute_operation(self, left: int, right: int, op: str) -> int:
        """Hidden state operation."""
        if op == "+":
            return left + right
        elif op == "-":
            return left - right
        elif op == "*":
            return left * right
        elif op == "/" and right != 0:
            return left // right
        return left

    def get_display(self) -> str:
        """Observation: current display value."""
        return self.display

    def get_history(self) -> List[str]:
        """Observation: operation history."""
        return self.history.copy()

    def assert_display(self, expected: str) -> bool:
        """Preference: assertion on display."""
        return self.display == expected

    def assert_history_length(self, max_length: int) -> bool:
        """Preference: constraint on history size."""
        return len(self.history) <= max_length


# Example usage
if __name__ == "__main__":
    calc = Calculator()

    # 5 + 3 = 8
    calc.input_digit(5)
    calc.input_operation("+")
    calc.input_digit(3)
    result = calc.equals()

    print(f"Result: {result}")
    print(f"History: {calc.get_history()}")

    # Assertions
    assert calc.assert_display("8"), "Display should be 8"
    assert calc.assert_history_length(10), "History should be short"
